src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -4255,23 +4184,33 @@ def _array_syntax_to_indexed(self, parent, loop_vars):
         :raises NotImplementedError: if array sections of differing ranks are
                                      found.
         '''
-        assigns = parent.walk(Assignment)
-        # Check that the LHS of any assignment uses array notation.
-        # Note that this will prevent any WHERE blocks that contain scalar
-        # assignments from being handled but is a necessary limitation until
-        # #717 is done and we interrogate the type of each symbol.
-        for assign in assigns:
-            _ = self._array_notation_rank(assign.lhs)
-
-        # TODO #717 if the supplied code accidentally omits array
-        # notation for an array reference on the RHS then we will
-        # identify it as a scalar and the code produced from the
-        # PSyIR (using e.g. the Fortran backend) will not
-        # compile. We need to implement robust identification of the
-        # types of all symbols in the PSyIR fragment.
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.transformations import (
+                Reference2ArrayRangeTrans, TransformationError)
+        # Convert References to arrays to use the array range notation unless
+        # they have an IntrinsicCall parent.
+        for ref in parent.walk(Reference):
+            if isinstance(ref.symbol.interface, ImportInterface):
+                raise NotImplementedError(
+                        "PSyclone doesn't yet support reference to imported "
+                        "symbols inside WHERE clauses.")
+            intrinsic_ancestor = ref.ancestor(IntrinsicCall)
+            if (isinstance(ref.symbol, DataSymbol) and
+                    not intrinsic_ancestor):
+                try:
+                    Reference2ArrayRangeTrans().apply(ref)
+                except TransformationError:
+                    pass
+            elif (isinstance(ref.symbol, DataSymbol) and intrinsic_ancestor
```

Comment:
What happens if it's not a DataSymbol? Some are fine, like a RoutineSymbol, but some may need to raise a NotImplementedError, e.g. like a baseclass Symbol (which can happen with unqualified imports).

Reply:
I assumed it would already be handled by the frontend to result in a NotImplementedError if we weren't sure, e.g. `ImportInterface` is already blocked by line 4422 (`if isinstance(ref.symbol.interface, ImportInterface):`). I didn't feel like we needed to explicitly handle those during this function as they'd fail with a more accurate message later/earlier (depending on when we're calling this).

Reply:
I see you added UnresovledInterface in the commit (which should be the same as symbols that use the baseclass as they could not be found)  and the performance of NEMO did not degrade. This is safer than before so I would keep this change. You will need to fix the new failing tests (probably you will need to convert some imported symbols to local?) and would be good to add a test with an unqualified import producing the NotImplementedError/Codeblock (even if the coverage says its the same line as the ImportInterface check).

Reply:
I think some of the failing tests were testing "what happens if unqualified import" so I'll convert those into these tests probably.

Reply:
Done